txscript/opcode.go

```diff
@@ -1850,7 +1850,17 @@ func opcodeCheckSig(op *parsedOpcode, vm *Engine) error {
 		return nil
 	}
 
-	ok := signature.Verify(hash, pubKey)
+	var ok bool
+	if vm.sigCache != nil {
+		ok = vm.sigCache.Exists(hash, signature, pubKey)
+		if !ok && signature.Verify(hash, pubKey) {
+			vm.sigCache.Add(hash, signature, pubKey)
+			ok = true
+		}
+	} else {
+		ok = signature.Verify(hash, pubKey)
+	}
```

Comment:
also I think this could would be a little more readable as:

```Go
valid := false
if vm.sigCache != nil {
    valid = vm.sigCache.Exists(...)
}
if !valid {
    valid = signature.Verify(...)
    if valid {
        vm.sigCache.Add(...)
    }
}
```